src/Microsoft.AspNet.Server.Kestrel/Http/SocketInput.cs

```diff
@@ -194,11 +194,13 @@ public void OnCompleted(Action continuation)
             }
             else if (awaitableState == _awaitableIsCompleted)
             {
-                ThreadPool.QueueUserWorkItem((o) => ((Action)o)(), continuation);
+                // Read already completed, continue on calling thread
+                continuation();
```

Comment:
Why are we trying hard to avoid queuing here? This should be rare: assuming this call is part of an await, we just checked to see if it already completed, so it would have needed to complete between that check and this one immediately after. Are you seeing this case a high percentage of the time?